internal/pkg/client/cache/dir.go

```diff
@@ -3,93 +3,308 @@
+	}
+
+	// If the environment variable is not set, we use the default cache.
+	usr, err := user.Current()
 	if err != nil {
-		sylog.Fatalf("Unable to get abs filepath: %v", err)
+		return "", fmt.Errorf("couldn't determine user home directory: %s", err)
+	}
+	basedir = path.Join(usr.HomeDir, basedirDefault)
+
+	return basedir, nil
+}
+
+// getCacheRoot figures out what the root directory is.
+func getCacheRoot(basedir string) string {
+	// Note: basedir and root are different and described earlier.
+	// The DirEnv environment variable can be used to set the basedir (or it
+	// is set to a default location) and the root is the actual cache directory
+	// within basedir, which we can safely delete (it is only supposed to contain
+	// files and directories that we create).
+	return path.Join(basedir, CacheDir)
+}
+
+// updateCacheSubdir update/create a sub-cache (directory) within the cache,
+// for example, the 'shub' cache.
+func updateCacheSubdir(c *SingularityCache, subdir string) (string, error) {
+	// This function may act on an cache object that is not fully initialized
+	// so it is not a method on a SingularityCache but rather an independent
+	// function.
+	// Because the cache object may not be initialized, we do NOT check its validity
+	if c == nil {
+		return "", fmt.Errorf("invalid cache handle")
+	}
+
+	// If the subdir is empty, it will lead to new collision since it would
+	// succeed but point at the cache's root
+	if subdir == "" {
+		return "", fmt.Errorf("invalid parameter")
+	}
+
+	absdir, err := filepath.Abs(filepath.Join(c.rootDir, subdir))
+	if err != nil {
+		return "", fmt.Errorf("unable to get abs filepath: %v", err)
 	}
 
 	if err := initCacheDir(absdir); err != nil {
-		sylog.Fatalf("Unable to initialize caching directory: %v", err)
+		return "", fmt.Errorf("unable to initialize caching directory: %v", err)
 	}
 
 	sylog.Debugf("Caching directory set to %s", absdir)
-	return absdir
+	return absdir, nil
 }
 
+// initCacheDir initializes a sub-cache within a cache, e.g., the shub
+// sub-cache.
 func initCacheDir(dir string) error {
-	if _, err := os.Stat(dir); os.IsNotExist(err) {
+	fInfo, err := os.Stat(dir)
+	switch {
+	case os.IsNotExist(err):
+		// The directory does not exist, we create it
 		sylog.Debugf("Creating cache directory: %s", dir)
 		if err := fs.MkdirAll(dir, 0755); err != nil {
 			return fmt.Errorf("couldn't create cache directory %v: %v", dir, err)
 		}
-	} else if err != nil {
+		return nil
+	case err != nil:
+		// A actual error occurred
 		return fmt.Errorf("unable to stat %s: %s", dir, err)
+	case !fInfo.IsDir():
+		// This is actually not a directory
+		return fmt.Errorf("%s is not a directory", dir)
+	default:
+		// The directory exists
+		return nil
 	}
+}
 
-	return nil
+// checkImageHash checks whether the sha256 hash of an image in the cache
+// matches the sum passed in.
+func checkImageHash(path, sum string) bool {
+	cacheSum, err := client.ImageHash(path)
+	if err != nil {
```

Comment:
Notice how this is checking booleans to return booleans. This means the whole thing can be written as:

    return err == nil && sum == cacheSum

Reply:
fixed